File: agent/log_exporter.py
"""Best-effort duplication of process output to a generic syslog endpoint."""

from __future__ import annotations

from collections.abc import Callable, Mapping
from dataclasses import dataclass, field
import json
import logging
import os
import queue
import re
import socket
import sys
import threading
from typing import Any, Literal, TextIO, cast


MAX_DATAGRAM_BYTES = 8_192
MAX_INCOMPLETE_BYTES = 65_536
INCOMPLETE_FLUSH_BYTES = 8_000
LogComponent = Literal["runtime", "agent"]
StreamKind = Literal["stdout", "stderr"]


@dataclass(frozen=True)
class LogExporterHandle:
    """Lifecycle handle returned by :func:`install_log_exporter`."""

    enabled: bool
    _close: Callable[[], None] = field(default=lambda: None, repr=False)

    def close(self) -> None:
        """Restore wrapped streams and release exporter resources."""
        self._close()


DISABLED_HANDLE = LogExporterHandle(enabled=False)


@dataclass(frozen=True)
class _Configuration:
    host: str
    port: int
    component: LogComponent


def _diagnose(stderr: TextIO, detail: str) -> None:
    try:
        stderr.write(f"[opentag] log exporter disabled: {detail}\n")
    except Exception:
        pass


def _configuration(
    environment: Mapping[str, str], stderr: TextIO
) -> _Configuration | None:
    exporter = environment.get("OPENTAG_LOG_EXPORTER", "none")
    if exporter == "none":
        return None
    if exporter != "syslog":
        _diagnose(stderr, f'invalid OPENTAG_LOG_EXPORTER "{exporter}"')
        return None

    host = environment.get("OPENTAG_LOG_EXPORTER_HOST", "").strip()
    raw_port = environment.get("OPENTAG_LOG_EXPORTER_PORT", "")
    component = environment.get("OPENTAG_LOG_COMPONENT", "")
    try:
        port = int(raw_port)
        valid_port = 1 <= port <= 65535
    except ValueError:
        port = 0
        valid_port = False
    if not host or not valid_port or component not in {"runtime", "agent"}:
        _diagnose(
            stderr,
            "syslog requires a host, a valid port, and component runtime|agent",
        )
        return None
    return _Configuration(
        host=host,
        port=port,
        component=cast(LogComponent, component),
    )


_TEXT_LEVEL = re.compile(
    r"^\s*(?:\[\s*)?"
    r"(emerg(?:ency)?|alert|fatal|crit(?:ical)?|err(?:or)?|warn(?:ing)?|"
    r"notice|info(?:rmation)?|debug|trace)"
    r"(?:\s*\])?(?=\s|:|-|$)",
    re.IGNORECASE,
)


def _named_severity(value: str) -> int | None:
    normalized = value.strip().lower()
    if normalized in {"emerg", "emergency"}:
        return 0
    if normalized == "alert":
        return 1
    if normalized in {"fatal", "critical", "crit"}:
        return 2
    if normalized in {"error", "err"}:
        return 3
    if normalized in {"warn", "warning"}:
        return 4
    if normalized == "notice":
        return 5
    if normalized in {"info", "information"}:
        return 6
    if normalized in {"trace", "debug"}:
        return 7
    return None


def _numeric_severity(value: float) -> int | None:
    if value >= 60:
        return 2
    if value >= 50:
        return 3
    if value >= 40:
        return 4
    if value >= 30:
        return 6
    if value >= 0:
        return 7
    return None


def _infer_severity(message: str, kind: StreamKind) -> int:
    try:
        parsed = json.loads(message)
        if isinstance(parsed, dict):
            for key in ("level", "severity", "severityText"):
                value = parsed.get(key)
                inferred = None
                if isinstance(value, str):
                    inferred = _named_severity(value)
                elif isinstance(value, (int, float)) and not isinstance(value, bool):
                    inferred = _numeric_severity(value)
                if inferred is not None:
                    return inferred
    except (TypeError, ValueError):
        pass

    match = _TEXT_LEVEL.match(message)
    if match:
        inferred = _named_severity(match.group(1))
        if inferred is not None:
            return inferred
    return 6 if kind == "stdout" else 3


def _take_utf8_prefix(value: str, max_bytes: int) -> tuple[str, str]:
    bytes_used = 0
    end = 0
    for character in value:
        character_bytes = len(character.encode())
        if bytes_used + character_bytes > max_bytes:
            break
        bytes_used += character_bytes
        end += len(character)
    return value[:end], value[end:]


def _frames(
    message: str, kind: StreamKind, component: LogComponent
) -> list[bytes]:
    severity = _infer_severity(message, kind)
    priority = 16 * 8 + severity
    header = f"<{priority}>1 - - opentag-{component} - - - "
    payload_bytes = MAX_DATAGRAM_BYTES - len(header.encode())
    if not message:
        return [header.encode()]

    result = []
    remaining = message
    while remaining:
        chunk, remaining = _take_utf8_prefix(remaining, payload_bytes)
        result.append((header + chunk).encode())
    return result


class _Sender:
    def __init__(self, datagram_socket: Any, configuration: _Configuration):
        self._socket = datagram_socket
        self._configuration = configuration
        self._queue: queue.Queue[bytes | None] = queue.Queue(maxsize=1024)
        self._closed = False
        self._thread = threading.Thread(
            target=self._run, name="opentag-log-exporter", daemon=True
        )
        self._thread.start()

    def send(self, payload: bytes) -> None:
        if self._closed:
            return
        try:
            self._queue.put_nowait(payload)
        except queue.Full:
            pass

    def _run(self) -> None:
        while True:
            payload = self._queue.get()
            if payload is None:
                return
            try:
                self._socket.sendto(
                    payload,
                    (self._configuration.host, self._configuration.port),
                )
            except Exception:
                pass

    def close(self) -> None:
        if self._closed:
            return
        self._closed = True
        try:
            self._queue.put_nowait(None)
        except queue.Full:
            pass
        try:
            self._socket.close()
        except Exception:
            pass


class _LineForwarder:
    def __init__(
        self, kind: StreamKind, component: LogComponent, sender: _Sender
    ):
        self._kind = kind
        self._component = component
        self._sender = sender
        self._buffer = ""

    def write(self, value: str) -> None:
        self._buffer += value
        while "\n" in self._buffer:
            line, self._buffer = self._buffer.split("\n", 1)
            for frame in _frames(line, self._kind, self._component):
                self._sender.send(frame)
        while len(self._buffer.encode()) > MAX_INCOMPLETE_BYTES:
            chunk, self._buffer = _take_utf8_prefix(
                self._buffer, INCOMPLETE_FLUSH_BYTES
            )
            for frame in _frames(chunk, self._kind, self._component):
                self._sender.send(frame)


class _StreamProxy:
    def __init__(self, stream: TextIO, forwarder: _LineForwarder):
        self._stream = stream
        self._forwarder = forwarder

    def write(self, value: str) -> int:
        result = self._stream.write(value)
        try:
            self._forwarder.write(value)
        except Exception:
            pass
        return result

    def flush(self) -> None:
        return self._stream.flush()

    def __getattr__(self, name: str) -> Any:
        return getattr(self._stream, name)


def _rebind_logging_handlers(
    original_stdout: TextIO,
    original_stderr: TextIO,
    stdout_proxy: _StreamProxy,
    stderr_proxy: _StreamProxy,
) -> list[tuple[logging.StreamHandler, TextIO]]:
    loggers = [logging.getLogger()]
    loggers.extend(
        logger
        for logger in logging.Logger.manager.loggerDict.values()
        if isinstance(logger, logging.Logger)
    )
    rebound = []
    seen = set()
    for logger in loggers:
        for handler in logger.handlers:
            if not isinstance(handler, logging.StreamHandler) or id(handler) in seen:
                continue
            seen.add(id(handler))
            if handler.stream is original_stdout:
                try:
                    handler.setStream(stdout_proxy)
                    rebound.append((handler, original_stdout))
                except Exception:
                    pass
            elif handler.stream is original_stderr:
                try:
                    handler.setStream(stderr_proxy)
                    rebound.append((handler, original_stderr))
                except Exception:
                    pass
    return rebound


def install_log_exporter(
    *,
    env: Mapping[str, str] | None = None,
    stdout: TextIO | None = None,
    stderr: TextIO | None = None,
    socket_factory: Callable[[int, int], Any] = socket.socket,
) -> LogExporterHandle:
    """Install log forwarding when explicitly enabled by the environment."""
    environment = os.environ if env is None else env
    original_stdout = sys.stdout if stdout is None else stdout
    original_stderr = sys.stderr if stderr is None else stderr
    configuration = _configuration(environment, original_stderr)
    if configuration is None:
        return DISABLED_HANDLE

    try:
        datagram_socket = socket_factory(socket.AF_INET6, socket.SOCK_DGRAM)
        datagram_socket.setblocking(False)
    except Exception:
        _diagnose(original_stderr, "could not create the IPv6 UDP socket")
        return DISABLED_HANDLE

    try:
        sender = _Sender(datagram_socket, configuration)
    except Exception:
        try:
            datagram_socket.close()
        except Exception:
            pass
        _diagnose(original_stderr, "could not start the asynchronous sender")
        return DISABLED_HANDLE
    stdout_proxy = _StreamProxy(
        original_stdout,
        _LineForwarder("stdout", configuration.component, sender),
    )
    stderr_proxy = _StreamProxy(
        original_stderr,
        _LineForwarder("stderr", configuration.component, sender),
    )
    if sys.stdout is original_stdout:
        sys.stdout = stdout_proxy  # type: ignore[assignment]
    if sys.stderr is original_stderr:
        sys.stderr = stderr_proxy  # type: ignore[assignment]
    rebound_handlers = _rebind_logging_handlers(
        original_stdout,
        original_stderr,
        stdout_proxy,
        stderr_proxy,
    )
    closed = False

    def close() -> None:
        nonlocal closed
        if closed:
            return
        closed = True
        if sys.stdout is stdout_proxy:
            sys.stdout = original_stdout
        if sys.stderr is stderr_proxy:
            sys.stderr = original_stderr
        for handler, original_stream in rebound_handlers:
            if handler.stream in {stdout_proxy, stderr_proxy}:
                try:
                    handler.setStream(original_stream)
                except Exception:
                    pass
        sender.close()

    return LogExporterHandle(enabled=True, _close=close)

File: app/log-exporter.ts
import { createSocket as createDatagramSocket, type Socket } from "node:dgram";
import type { Writable } from "node:stream";
import { StringDecoder } from "node:string_decoder";
import {
  readLogExporterEnvironment,
  type LogComponent,
} from "./env.js";

type StreamKind = "stdout" | "stderr";

const MAX_DATAGRAM_BYTES = 8_192;
const MAX_INCOMPLETE_BYTES = 65_536;
const INCOMPLETE_FLUSH_BYTES = 8_000;

export interface LogExporterHandle {
  readonly enabled: boolean;
  close(): void;
}

export interface InstallLogExporterOptions {
  env?: NodeJS.ProcessEnv;
  stdout?: Writable;
  stderr?: Writable;
  createSocket?: (type: "udp6") => Socket;
}

const DISABLED_HANDLE: LogExporterHandle = {
  enabled: false,
  close() {},
};

function reportConfigurationError(stderr: Writable, detail: string): void {
  try {
    stderr.write(`[opentag] log exporter disabled: ${detail}\n`);
  } catch {
    // Logging configuration must never affect application startup.
  }
}

function namedSeverity(value: string): number | undefined {
  switch (value.trim().toLowerCase()) {
    case "emerg":
    case "emergency":
      return 0;
    case "alert":
      return 1;
    case "fatal":
    case "critical":
    case "crit":
      return 2;
    case "error":
    case "err":
      return 3;
    case "warn":
    case "warning":
      return 4;
    case "notice":
      return 5;
    case "info":
    case "information":
      return 6;
    case "trace":
    case "debug":
      return 7;
    default:
      return undefined;
  }
}

function numericSeverity(value: number): number | undefined {
  if (!Number.isFinite(value)) return undefined;
  if (value >= 60) return 2;
  if (value >= 50) return 3;
  if (value >= 40) return 4;
  if (value >= 30) return 6;
  if (value >= 0) return 7;
  return undefined;
}

function inferSeverity(message: string, kind: StreamKind): number {
  try {
    const parsed = JSON.parse(message) as Record<string, unknown>;
    if (parsed && typeof parsed === "object" && !Array.isArray(parsed)) {
      for (const key of ["level", "severity", "severityText"]) {
        const value = parsed[key];
        const inferred =
          typeof value === "string"
            ? namedSeverity(value)
            : typeof value === "number"
              ? numericSeverity(value)
              : undefined;
        if (inferred !== undefined) return inferred;
      }
    }
  } catch {
    // Most log lines are plain text, not JSON.
  }

  const prefix = message.match(
    /^\s*(?:\[\s*)?(emerg(?:ency)?|alert|fatal|crit(?:ical)?|err(?:or)?|warn(?:ing)?|notice|info(?:rmation)?|debug|trace)(?:\s*\])?(?=\s|:|-|$)/i,
  )?.[1];
  return (prefix ? namedSeverity(prefix) : undefined) ??
    (kind === "stdout" ? 6 : 3);
}

function takeUtf8Prefix(value: string, maxBytes: number): [string, string] {
  let bytes = 0;
  let end = 0;
  for (const character of value) {
    const characterBytes = Buffer.byteLength(character);
    if (bytes + characterBytes > maxBytes) break;
    bytes += characterBytes;
    end += character.length;
  }
  return [value.slice(0, end), value.slice(end)];
}

function syslogFrames(
  message: string,
  kind: StreamKind,
  component: LogComponent,
): Buffer[] {
  const severity = inferSeverity(message, kind);
  const priority = 16 * 8 + severity;
  const header = `<${priority}>1 - - opentag-${component} - - - `;
  const payloadBytes = MAX_DATAGRAM_BYTES - Buffer.byteLength(header);
  if (message.length === 0) return [Buffer.from(header, "utf8")];

  const frames: Buffer[] = [];
  let remaining = message;
  while (remaining.length > 0) {
    const [chunk, rest] = takeUtf8Prefix(remaining, payloadBytes);
    frames.push(Buffer.from(header + chunk, "utf8"));
    remaining = rest;
  }
  return frames;
}

function patchStream(
  stream: Writable,
  kind: StreamKind,
  forward: (message: string, kind: StreamKind) => void,
): () => void {
  const originalWrite = stream.write;
  let buffered = "";
  let decoder = new StringDecoder("utf8");

  const patchedWrite = function (this: Writable, ...args: unknown[]): boolean {
    const result = Reflect.apply(originalWrite, stream, args) as boolean;
    try {
      const chunk = args[0];
      if (typeof chunk === "string") {
        buffered += decoder.end() + chunk;
        decoder = new StringDecoder("utf8");
      } else if (chunk instanceof Uint8Array) {
        buffered += decoder.write(Buffer.from(chunk));
      }

      let newline = buffered.indexOf("\n");
      while (newline >= 0) {
        forward(buffered.slice(0, newline), kind);
        buffered = buffered.slice(newline + 1);
        newline = buffered.indexOf("\n");
      }

      while (Buffer.byteLength(buffered) > MAX_INCOMPLETE_BYTES) {
        const [chunkToForward, rest] = takeUtf8Prefix(
          buffered,
          INCOMPLETE_FLUSH_BYTES,
        );
        forward(chunkToForward, kind);
        buffered = rest;
      }
    } catch {
      // Mirroring is best-effort and cannot change stream behavior.
    }
    return result;
  };

  stream.write = patchedWrite as typeof stream.write;
  return () => {
    if (stream.write === patchedWrite) stream.write = originalWrite;
  };
}

export function installLogExporter(
  options: InstallLogExporterOptions = {},
): LogExporterHandle {
  const env = options.env ?? process.env;
  const stdout = options.stdout ?? process.stdout;
  const stderr = options.stderr ?? process.stderr;
  const configuration = readLogExporterEnvironment(env);
  if (configuration.exporter === "none") return DISABLED_HANDLE;
  if (configuration.exporter === "invalid") {
    reportConfigurationError(stderr, configuration.detail);
    return DISABLED_HANDLE;
  }

  let socket: Socket | undefined;
  try {
    socket = (options.createSocket ?? createDatagramSocket)("udp6");
    socket.on("error", () => {});
    socket.unref();
  } catch {
    try {
      socket?.close();
    } catch {
      // Ignore cleanup failures on a partially initialized socket.
    }
    reportConfigurationError(stderr, "could not create the IPv6 UDP socket");
    return DISABLED_HANDLE;
  }

  const activeSocket = socket;

  const forward = (message: string, kind: StreamKind): void => {
    for (const frame of syslogFrames(message, kind, configuration.component)) {
      try {
        activeSocket.send(
          frame,
          configuration.port,
          configuration.host,
          () => {},
        );
      } catch {
        // Network failures are intentionally silent and fail open.
      }
    }
  };
  let restoreStdout = (): void => {};
  let restoreStderr = (): void => {};
  try {
    restoreStdout = patchStream(stdout, "stdout", forward);
    restoreStderr = patchStream(stderr, "stderr", forward);
  } catch {
    restoreStdout();
    restoreStderr();
    try {
      activeSocket.close();
    } catch {
      // Ignore cleanup failures on a partially initialized exporter.
    }
    reportConfigurationError(stderr, "could not wrap process output");
    return DISABLED_HANDLE;
  }
  let closed = false;

  return {
    enabled: true,
    close() {
      if (closed) return;
      closed = true;
      restoreStdout();
      restoreStderr();
      try {
        activeSocket.close();
      } catch {
        // Closing a failed or already-closed socket is harmless.
      }
    },
  };
}
